ncdr/importers/column.py

```diff
@@ -1,11 +1,34 @@
 import collections
 import csv
 
-from django.db import transaction
-
 from ..models import Column, DataElement, Table
 
 
+def get_data_elementLUT(rows):
+    """
+    Build a lookup table of DataElement instances keyed by their name
+
+    This replaces a get_or_create in the outer loop body below.  Each iteration
+    would get call the get_or_create for the Column instantiation to use.
+    During the move to import CSVs via webform it was noted importing took ~90s
+    on a local machine.  However when that same import function, import_data at
+    the time of writing, was called by an RQ worker it ran in ~10s.  After much
+    digging it was discovered that the aggregate time cost for the
+    get_or_creates calls was the cause, again only when called directly from
+    the view.  Rather than waste more time building an extension on the rabbit
+    hole the sensible decision to use a precomputed lookup table was taken.
+
+    This function builds that table, creating any missing DataElements in the
+    process.
+    """
+    existing_de_names = set(DataElement.objects.values_list("name", flat=True))
```

Comment:
we could do existing_de_names = set(DataElement.objects.all().filter(name__in= csv_de_names).values_list("name", flat=True))

(bit bike sheddy...)

Reply:
You still have to compute `missing_de_names` though so you've asked the database to do a slightly harder query (by adding a filter clause) without saving the CPU any time when it calculates the set difference.